### PR TITLE
DAOS-5962 bio/dmg: Call bio_dev_list() from list-devices cmd

### PR DESCRIPTION
**Pull request by sydneyvanda**
Follow-on patch from DAOS-5905 (PR 3719).
Update dmg storage query list-devices command to list all BIO
devices, which include all unplugged and new devices in addition
to all devices in SMD that are currently in-use by DAOS.

Update BIO device states to be returned to admin via the
dmg storage query list-devices command. Previously only the
persistent SMD device states were returned by this command
(ie NORMAL or FAULTY). NEW BIO device states are now:
NORMAL = device in-use by DAOS
EVICTED = device has been manually set to FAULTY
UNPLUGGED = device is unplugged
NEW = device is plugged in but not yet in-use

Signed-off-by: Sydney Vanda <sydney.m.vanda@intel.com>

### CODE REVIEW

File: src/include/daos_srv/bio.h [hunk 1]
```diff
@@ -310,6 +310,47 @@ bio_sgl_holes(struct bio_sglist *bsgl)
 	return result;
 }
 
+enum bio_dev_state {
+	BIO_DEV_NORMAL	= 0, /* fully functional and in-use */
+	BIO_DEV_FAULTY,      /* evicted device */
+	BIO_DEV_OUT,         /* unplugged device */
+	BIO_DEV_NEW,         /* new device not currently in-use */
+};
```
Comment (NiuYawei) [inline src/include/daos_srv/bio.h:318]: These are all "displayed" state, I don't think it's required in "bio.h". I don't think it's quite necessary to define these displayed state, if it's useful for control plane, please move the definition to control plane.

File: src/bio/bio_device.c [hunk 2]
```diff
@@ -456,3 +455,140 @@ bio_replace_dev(struct bio_xs_context *xs_ctxt, uuid_t old_dev_id,
...
+		b_info = alloc_dev_info(d_bdev->bb_uuid, s_info);
+		if (b_info == NULL) {
+			D_ERROR("Failed to allocate device info\n");
+			rc = -DER_NOMEM;
+			goto out;
+		}
+		if (!d_bdev->bb_removed)
+			b_info->bdi_flags |= NVME_DEV_FL_PLUGGED;
+
+		/* set BIO device state based on flags */
```
Comment (NiuYawei) [inline src/bio/bio_device.c:554]: I don't think we should convert the flags into the displayed state here.

These flags are introduced for passing device state internally. With these three flags, we can describe all the device state combinations accurately. (for example, we can tell if an unplugged device was marked as faulty by these flags, but there is no such displayed state) 

The displayed state should only be used by control plane. I don't have strong opinion on how to display device state to end user, but for internal device state description, it should be done by these flags.

File: src/bio/bio_monitor.c [hunk 3]
```diff
@@ -141,6 +154,56 @@ bio_get_bs_state(int *bs_state, struct bio_xs_context *xs)
...
+	*n_dev_list = bdl.dev_list_cnt;
+
+out:
+	rc = ABT_eventual_free(&bdl.eventual);
+	if (rc != ABT_SUCCESS)
+		rc = dss_abterr2der(rc);
+
+	return rc;
+}
+
```
Comment (NiuYawei) [inline src/bio/bio_monitor.c:206]: Above two functions can be removed if you use dss_ult_execute().

File: src/mgmt/srv_query.c [hunk 4]
```diff
@@ -224,32 +224,73 @@ ds_mgmt_bio_health_query(struct mgmt_bio_health *mbh, uuid_t dev_uuid,
...
+
+	bxc = info->dmi_nvme_ctxt;
+	if (bxc == NULL) {
+		D_ERROR("BIO NVMe context not initialized for xs:%d, tgt:%d\n",
+			info->dmi_xs_id, info->dmi_tgt_id);
+		return;
+	}
+
+	rc = bio_get_dev_list(&list_devs_info->dev_list,
+			      &list_devs_info->dev_list_cnt, bxc);
```
Comment (NiuYawei) [inline src/mgmt/srv_query.c:250]: Call bio_dev_list() directly here if you use dss_ult_execute().

File: src/mgmt/srv_query.c [hunk 5]
```diff
@@ -224,32 +224,73 @@ ds_mgmt_bio_health_query(struct mgmt_bio_health *mbh, uuid_t dev_uuid,
...
-	struct smd_dev_info	*dev_info = NULL, *tmp;
-	d_list_t		 dev_list;
-	int			 dev_list_cnt = 0;
-	int			 buflen = 10;
-	int			 i = 0, j;
-	int			 rc = 0;
-
-	D_DEBUG(DB_MGMT, "Querying SMD device list\n");
+	struct bio_dev_info	   *dev_info = NULL, *tmp;
+	struct bio_list_devs_info  *list_devs_info = NULL;
```
Comment (NiuYawei) [inline src/mgmt/srv_query.c:261]: This can be put on stack.

File: src/bio/bio_device.c (hunk 2 above)
Comment (sydneyvanda) [reply src/bio/bio_device.c:554]: I understand your argument here, trying to figure out the best place to instead determine the states based on the flags in the control plane. @tanabarr any suggestions?

File: src/mgmt/srv_query.c [hunk 6]
```diff
@@ -224,32 +224,73 @@ ds_mgmt_bio_health_query(struct mgmt_bio_health *mbh, uuid_t dev_uuid,
...
+	if (list_devs_info == NULL) {
+		D_ERROR("Failed to allocate bio list devs info struct");
+		return -DER_NOMEM;
+	}
+	D_INIT_LIST_HEAD(&list_devs_info->dev_list);
 
-	D_INIT_LIST_HEAD(&dev_list);
-	rc = smd_dev_list(&dev_list, &dev_list_cnt);
+	rc = dss_ult_create(bio_query_dev_list, list_devs_info, DSS_ULT_GC, 0,
+			    0, &thread);
```
[the remark this answers is not in the document]
Comment (sydneyvanda) [reply src/mgmt/srv_query.c:277]: ok I updated to use dss_ult_execute() on tgtid 0, however I am still unsure how to guarantee that the ULT will be created on the init xstream w/o using the spdk_thread_send_msg(init_thread()) ?

File: src/mgmt/srv_query.c (hunk 5 above)
Comment (sydneyvanda) [reply src/mgmt/srv_query.c:261]: done

File: src/mgmt/srv_query.c (hunk 4 above)
Comment (sydneyvanda) [reply src/mgmt/srv_query.c:250]: done

File: src/bio/bio_monitor.c (hunk 3 above)
Comment (sydneyvanda) [reply src/bio/bio_monitor.c:206]: done

File: src/bio/bio_device.c (hunk 2 above)
Comment (tanabarr) [reply src/bio/bio_device.c:556]: you can pass bdi_flags back in the drpc response and use bitwise operations with imported flag definitions e.g. NVME_DEV_FL_FAULTY to display a state just like you are doing below but in the control-plane server, then just pass strings back to DMG. that makes sense . or did I misunderstand your question?

File: src/include/daos_srv/bio.h (hunk 1 above)
Comment (tanabarr) [reply src/include/daos_srv/bio.h:318]: these constants could be defined in control/server/storage/types.go {look at ScmState* for an example) then they will be able to be imported from both server and dmg so you can set state based on dRPC response in the server (bdi_flags uint8) serialize state in grpc response and unpack in dmg, adding a stringer interface implementation to the state type in types.go will then enable you to print a string representation of the state received in dmg

File: src/include/daos_srv/bio.h [hunk 7]
```diff
@@ -637,4 +678,14 @@ void bio_log_csum_err(struct bio_xs_context *b, int tgt_id);
...
 bool bio_need_nvme_poll(struct bio_xs_context *xs);
+
+/*
+ * Convert device state to human-readable string
+ *
+ * \param [IN]  state	Device state
+ *
+ * \return		Static string representing enum value
+ */
+char *bio_dev_state_enum_to_str(enum bio_dev_state state);
```
Comment (tanabarr) [inline src/include/daos_srv/bio.h:689]: as described above

File: src/include/daos_srv/control.h [hunk 8]
```diff
@@ -32,6 +32,15 @@
 #include <stdbool.h>
 #include <assert.h>
 
+enum {
+	/* Device is plugged */
+	NVME_DEV_FL_PLUGGED	= 0x1,
```
Comment (tanabarr) [inline src/include/daos_srv/control.h:37]: these can be used to decode bdi_flags in the drpc list-devices caller in src/control/server package

**Review by tanabarr — COMMENTED** [review]
so you could remove the Logic to convert to strings and just pass bdi_flags back over drpc to be decoded in the control-plane server which can import the bit-flags from control.h to decode the state

File: src/mgmt/srv_query.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/mgmt/srv_query.c]: Current implementation ensures init thread is on target 0, so the ULT is already on the init thread.

**Comment by sydneyvanda** [comment]
> you could remove the logic to convert to strings and just pass bdi_flags back over drpc to be decoded in the control-plane server which can import the bit-flags from control.h to decode the state

I ended up moving the definitions from bio.h and putting them in control.h. I also moved the logic to convert from flags -> device state into ds_mgmt_smd_list_devs() in src/mgmt/srv_query.c (this seemed simpler and I could keep the string state definition in the protobuf file already and not have to add a flags param).

Let me know if this will suffice @NiuYawei  and @tanabarr

**Review by NiuYawei — APPROVED** [review]

**Comment by NiuYawei** [comment]
> > you could remove the logic to convert to strings and just pass bdi_flags back over drpc to be decoded in the control-plane server which can import the bit-flags from control.h to decode the state
> 
> I ended up moving the definitions from bio.h and putting them in control.h. I also moved the logic to convert from flags -> device state into ds_mgmt_smd_list_devs() in src/mgmt/srv_query.c (this seemed simpler and I could keep the string state definition in the protobuf file already and not have to add a flags param).
> 
> Let me know if this will suffice @NiuYawei and @tanabarr

That looks good to me.

**Review by tanabarr — APPROVED** [review]
are you planning to add unit tests anywhere?

**Comment by sydneyvanda** [comment]
> are you planning to add unit tests anywhere?

Currently there is a python tests to test NORMAL and EVICTED (previously FAULTY, need to add a patch to update this test) states. I figure we could add a test for the OUT (unplugged) state, and later figure out a way to test the NEW device state as well, but probably roll both of those test cases into the suite of tests we need to add for the hotplug feature.  @ravalsam can we add these (OUT and NEW device states) to the list for tests for the hotplug feature?

File: src/tests/ftest/control/dmg_storage_query.py [hunk 9]
```diff
@@ -173,4 +173,4 @@ def test_dmg_storage_query_device_state(self):
             self.get_dmg_output("storage_set_faulty", uuid=dev[0])
 
         # Check that devices are in FAULTY state
```
Comment (tanabarr) [inline src/tests/ftest/control/dmg_storage_query.py:175]: update comment?

**Review by tanabarr — APPROVED** [review]

File: src/tests/ftest/control/dmg_storage_query.py (hunk 9 above)
Comment (sydneyvanda) [reply src/tests/ftest/control/dmg_storage_query.py:175]: the devices are in a faulty state, it just displays EVICTED, I guess that kind of confusing

**Review by NiuYawei — APPROVED** [review]

**Comment by sydneyvanda** [comment]
OSA test failures from DAOS-6019 and DAOS-6018, not sure what is causing rebuild test failures, all logs say is "Test setup failed"

**Review by mjmac — APPROVED** [review]

**Pull request merged**